### DISCUSSION TITLE
Attempting to read a bool vector from input

### DISCUSSION THREAD

**Discussion by rks171** [idaholab/moose#17859] [Q&A General] [answered]
I'm attempting to use the `InputParameters` class to read a bool type vector from the input file as follows:

```
       pars.set<std::vector<bool>>("param_name") = getParam<std::vector<bool>>("param_name");
```

When I try to run the code, I get the following error:

```
*** ERROR ***
unsupported type 'std::__1::vector<bool, std::__1::allocator<bool> >' for input parameter 'GlobalParams/PBModelParams/param_name'
```

Which makes me think a vector of `bool`s is not supported in the class?

If I change to using an `int` vector, it works.

**Comment by aeslaughter** [comment, marked as answer]
I think you are correct that this is not supported. I am not sure if this is intentional, but I imagine we don't support it because a std::vector<bool> is handled a bit [differently than other types](https://en.cppreference.com/w/cpp/container/vector_bool) that has caused us problems in the past.

@roystgnr Do you have any other info on this.

**Reply by roystgnr** [reply]
This one's not my fault!  Check out framework/src/parser/Parser.C - there's nothing in the Parameters or InputParameters class that would complain about a `vector<bool>`, but there's no option in the parser for instantiating one, and that `Parser::extractParams()` code looks to be the source of the error message above.

That might not be an intentional omission.  A lot of people do hate `vector<bool>` for not following the same sort of API as the other `vector` instantiations, but a `vector<unsigned char>` is a pretty popular substitute, and I don't see that in the Parser.C list either.

**Reply by permcody** [reply]
@rks171, You might follow Roy's suggestion and see if you can add the appropriate specialization in Parser.C. We would accept a tested contribution. 

There are several macros in there to do this. Vector of bools may cause issues since they are fundamentally different, but as Roy suggested, I would start with 'unsigned char' and see if that meets your needs. Honestly, you might just use unsigned short and move on. It's unlikely that you have enough Booleans in an input parameter that would have any effect on memory usage.

**Reply by rks171** [reply]
Thanks for the suggestions for this issue.  When you suggest to use `unsigned char`, does that mean the user would be inputting `t` or `f` for "true"  or "false" and then I would be checking to make sure that is what is entered in the parser side?  And I'm not sure if I understood @roystgnr correctly, but it sounds like `unsigned char` might not be supported either?  My current workaround is to just use `unsigned int` and have the user input 0/1.  It works, but it's just a little less straightforward than entering true/false or on/off.

**Reply by GiudGiud** [reply]
@rks171 did you go with unsigned int? Is it working for you

Let's have an issue about this. Any volunteer to post it

**Reply by rks171** [reply]
Yes I have been using `unsigned int` and it works fine as a workaround.
